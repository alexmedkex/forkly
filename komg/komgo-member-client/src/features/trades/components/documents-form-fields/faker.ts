import { mockCategories } from '../../../document-management/store/categories/mock-data'
import { mockProduct } from '../../../document-management/store/products/mock-data'
import { DocumentType } from '../../../document-management/store/types/document-type'

const fakeCommercialContractDocumentType: DocumentType = {
  product: mockProduct,
  category: mockCategories[0],
  name: 'Commercial Contract',
  id: 'commercialContract',
  fields: [],
  predefined: false
}

export const fakeCommercialContractDocumentTypes: DocumentType[] = [fakeCommercialContractDocumentType]

export const fakeCommercialContractDocumentWithoutExtension = {
  _id: 'document0',
  file: {
    size: 5000
  },
  name: 'documentName',
  fileName: 'documentFileName',
  typeId: 'commercialContract',
  fileType: 'application/something',
  type: {
    id: 'commercialContract'
  }
}

export const fakeCommercialContractDocumentWithExtension = {
  _id: 'document1',
  file: {
    size: 5000
  },
  name: 'documentName1',
  fileName: 'documentFileName1.pdf',
  typeId: 'commercialContract',
  fileType: 'application/pdf',
  type: {
    id: 'commercialContract'
  }
}
