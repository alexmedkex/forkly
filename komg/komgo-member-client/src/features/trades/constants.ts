import { ITradeDocument } from './store/types'
import {
  TradeSource,
  PriceUnit,
  PaymentTermsTimeUnit,
  PaymentTermsDayType,
  PaymentTermsWhen,
  PaymentTermsEventBase,
  Currency,
  ICargoBase,
  ITradeBase,
  CreditRequirements,
  CARGO_SCHEMA_VERSION,
  PARCEL_SCHEMA_VERSION,
  TRADE_SCHEMA_VERSION
} from '@komgo/types'
import { IParcelWithOverrides } from './components/cargo-form-fields/ParcelData'
import { v4 } from 'uuid'

export const NOTENOUGHINFO = 'Not enough information provided.'
export const UNRECOGNISED_STATUS = 'Status unrecognized'

export const LOC_STATUS = {
  TO_BE_FINANCED: 'TO_BE_FINANCED',
  SUBMITTED: 'SUBMITTED',
  RECEIVED: 'RECEIVED',
  ADVISED: 'ADVISED',
  AMENDMENT_REQUESTED: 'AMENDMENT_REQUESTED',
  AMENDMENT_UNDER_NEGOTIATION: 'AMENDMENT_UNDER_NEGOTIATION',
  AMENDMENT_ACCEPTED: 'AMENDMENT_ACCEPTED',
  PAID: 'PAID',
  REFUSED: 'REFUSED',
  EXPIRED: 'EXPIRED'
}

export enum LOC_CATEGORIES {
  TO_BE_FINANCED = 'TO_BE_FINANCED',
  IN_PROGRESS = 'IN_PROGRESS',
  ISSUED = 'ISSUED',
  CLOSED = 'CLOSED'
}

export enum TRADE_STATUS {
  ToBeFinanced = 'TO_BE_FINANCED',
  ToBeDiscounted = 'TO_BE_DISCOUNTED'
}

export const LOC_CATEGORY_COLOURS = {
  [LOC_CATEGORIES.TO_BE_FINANCED]: 'red',
  [LOC_CATEGORIES.IN_PROGRESS]: 'yellow',
  [LOC_CATEGORIES.ISSUED]: 'green'
}

export const READABLE_LOC_STATUS = {
  [LOC_CATEGORIES.TO_BE_FINANCED]: [LOC_STATUS.TO_BE_FINANCED],
  [LOC_CATEGORIES.IN_PROGRESS]: [LOC_STATUS.SUBMITTED],
  [LOC_CATEGORIES.ISSUED]: [
    LOC_STATUS.RECEIVED,
    LOC_STATUS.ADVISED,
    LOC_STATUS.AMENDMENT_ACCEPTED,
    LOC_STATUS.AMENDMENT_UNDER_NEGOTIATION,
    LOC_STATUS.AMENDMENT_ACCEPTED
  ],
  [LOC_CATEGORIES.CLOSED]: [LOC_STATUS.PAID, LOC_STATUS.REFUSED, LOC_STATUS.EXPIRED]
}

export const DESC = -1
export const ASC = 1
export const DESC_TEXT = 'descending'
export const ASC_TEXT = 'ascending'

export enum TradingRole {
  SELLER = 'seller',
  BUYER = 'buyer'
}

export const TRADING_ROLE_OPTIONS = {
  BUYER: TradingRole.BUYER.toUpperCase(),
  SELLER: TradingRole.SELLER.toUpperCase()
}

export const CREDIT_REQUIREMENT_LABEL = {
  DOCUMENTARY_LETTER_OF_CREDIT: 'Documentary LC',
  STANDBY_LETTER_OF_CREDIT: 'Standby LC',
  OPEN_CREDIT: 'Open Credit',
  OFFSET: 'Offset'
}

export const CREDIT_REQUIREMENT_DISPLAY_VALUES = {
  [CreditRequirements.DocumentaryLetterOfCredit]: [CREDIT_REQUIREMENT_LABEL.DOCUMENTARY_LETTER_OF_CREDIT],
  [CreditRequirements.StandbyLetterOfCredit]: [CREDIT_REQUIREMENT_LABEL.STANDBY_LETTER_OF_CREDIT],
  [CreditRequirements.OpenCredit]: [CREDIT_REQUIREMENT_LABEL.OPEN_CREDIT],
  [CreditRequirements.Offset]: [CREDIT_REQUIREMENT_LABEL.OFFSET]
}

export const tradeDefaultEmptyValues: Partial<ITradeBase> = {
  buyerEtrmId: '',
  sellerEtrmId: '',
  dealDate: '',
  seller: '',
  buyer: '',
  commodity: '',
  generalTermsAndConditions: '',
  deliveryPeriod: { startDate: '', endDate: '' },
  demurrageTerms: '',
  laytime: '',
  law: '',
  creditRequirement: '' as any,
  contractReference: '',
  priceFormula: ''
}

export const initialTradeData: ITradeBase = {
  version: TRADE_SCHEMA_VERSION.V2,
  ...(tradeDefaultEmptyValues as ITradeBase),
  source: TradeSource.Komgo,
  dealDate: '',
  priceOption: undefined,
  price: undefined,
  currency: Currency.USD,
  priceUnit: PriceUnit.BBL,
  quantity: undefined,
  minTolerance: undefined,
  maxTolerance: undefined,
  invoiceQuantity: undefined,
  paymentTermsOption: undefined,
  paymentTerms: undefined,
  deliveryTerms: undefined,
  deliveryLocation: undefined,
  requiredDocuments: []
}

export const initialCargoData: ICargoBase = {
  source: TradeSource.Komgo,
  grade: '',
  cargoId: '',
  parcels: [],
  sourceId: '',
  originOfGoods: '',
  quality: '',
  version: CARGO_SCHEMA_VERSION.V2
}

export const generateInitialParcelData = (): IParcelWithOverrides => {
  const uniqueKey = v4()

  // As we have all parcels stored in a list in formik, when a new parcel is created
  // we need a unique key to track which item in the list is being edited
  return { ...initialParcelData, uniqueKey }
}

export const initialParcelData: IParcelWithOverrides = {
  id: '',
  laycanPeriod: {
    startDate: '',
    endDate: ''
  },
  modeOfTransport: '',
  modeOfTransportOther: '',
  vesselIMO: 0,
  vesselName: '',
  loadingPort: '',
  dischargeArea: '',
  inspector: '',
  deemedBLDate: '',
  quantity: undefined,
  warehouseOperatorName: '',
  pipelineName: '',
  tankFarmOperatorName: '',
  destinationPlace: '',
  loadingPlace: '',
  version: PARCEL_SCHEMA_VERSION.V2
}

export const initialDocumentData: ITradeDocument = {
  name: '',
  typeId: '',
  file: null,
  fileName: '',
  fileType: ''
}

export const TRADE_EDITED_MESSAGE = 'Trade edited'
export const TRADE_CREATED_MESSAGE = 'New trade created'
export const TRADE_AND_CARGO_EDITED_MESSAGE = 'Trade and cargo edited'
export const TRADE_DELETED_MESSAGE = 'Trade deleted'

export const emptyDropdownItem = { text: '(none)', content: '(none)' }
