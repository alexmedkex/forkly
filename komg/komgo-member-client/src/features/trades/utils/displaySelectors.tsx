import { CreditRequirements, TRADE_SCHEMA, ITrade, Commodity, Status } from '@komgo/types'
import { List, Map } from 'immutable'
import { get } from 'lodash'
import Numeral from 'numeral'
import * as React from 'react'
import { Label as SemanticUILabel, Popup, SemanticCOLORS } from 'semantic-ui-react'
import { capitalize, sentenceCase } from '../../../utils/casings'
import { displayDate, isDate } from '../../../utils/date'
import { Counterparty } from '../../counterparties/store/types'
import { Roles } from '../../letter-of-credit-legacy/constants/roles'
import { ILetterOfCredit } from '../../letter-of-credit-legacy/types/ILetterOfCredit'
import { IMember } from '../../members/store/types'
import {
  CREDIT_REQUIREMENT_DISPLAY_VALUES,
  LOC_CATEGORY_COLOURS,
  NOTENOUGHINFO,
  READABLE_LOC_STATUS,
  TradingRole,
  UNRECOGNISED_STATUS,
  TRADE_STATUS
} from '../constants'
import { ITradeEnriched } from '../store/types'
import { findFieldFromSchema } from '../../../store/common/selectors/displaySelectors'
import { Strings } from '../../receivable-discounting-legacy/resources/strings'

export const displayYesOrNo = (trueOrFalse: boolean): string => (trueOrFalse ? 'Yes' : 'No')

export const displayPercentage = (percent: number = 0): string => {
  return Numeral(percent / 100).format('% 0.00')
}

export const displayQuantity = (quantity?: number, priceUnit?: string, nullValueDisplay = NOTENOUGHINFO): string => {
  if (!quantity || !priceUnit) {
    return nullValueDisplay
  }
  return `${formatQuantity(quantity)} ${priceUnit}`
}

export const formatQuantity = (quantity: number): string => {
  return `${Numeral(quantity).format('0,0')}`
}

export const displayTolerance = displayPercentage

export const displayCommodity = (commodity: string): string => {
  if (commodity === undefined) {
    // TODO: Remove check once properly implemented from VAKT (MM)
    return 'BFOET'
  }
  return Object.values(Commodity).includes(commodity) ? sentenceCase(commodity) : commodity
}

export const displayPrice = (
  price?: number,
  currency?: string,
  priceUnit?: string,
  nullValueDisplay = NOTENOUGHINFO
): string => {
  if (!price || !currency || !priceUnit) {
    return nullValueDisplay
  }
  return `${formatPrice(price)} ${currency}/${priceUnit}`
}

export const formatPrice = (price: number): string => {
  return `${Numeral(price).format('0,0.00')}`
}

export const displayEtrmId = (trade: ITradeEnriched, companyStaticId?: string, role?: string): string => {
  if (trade.buyer === companyStaticId || role === Roles.ISSUING_BANK) {
    return `${trade.buyerEtrmId}`
  } else if (trade.seller === companyStaticId || role === Roles.ADVISING_BANK) {
    return `${trade.sellerEtrmId}`
  }
  return NOTENOUGHINFO
}

export const getHumanReadableLCStatus = (status: string): string => {
  return (
    Object.keys(READABLE_LOC_STATUS).find(
      readableStatus =>
        (READABLE_LOC_STATUS as any)[readableStatus].find((matchingStatus: string) => matchingStatus === status) !==
        undefined
    ) || UNRECOGNISED_STATUS
  )
}

export const displayTradeStatus = (tradeStatus: string): string =>
  tradeStatus === TRADE_STATUS.ToBeDiscounted ? Strings.SellerTradeStatusDefault : sentenceCase(tradeStatus)

export const displayLCStatus = (status?: string): JSX.Element | undefined => {
  if (!status) {
    return
  }

  const humanReadableLCStatus = getHumanReadableLCStatus(status)
  return makeLCStatusPopup(humanReadableLCStatus, (LOC_CATEGORY_COLOURS as any)[
    humanReadableLCStatus
  ] as SemanticCOLORS)
}

const makeLCStatusPopup = (text: string, color?: SemanticCOLORS): JSX.Element => {
  return (
    <Popup color={color} trigger={<SemanticUILabel circular={true} color={color} basic={!color && true} />}>
      {capitalize(text.toLocaleLowerCase().replace(/_/g, ' '))}
    </Popup>
  )
}

export const selectVisibleTrades = (trades: Map<string, any>, tradeIds: List<string>): ITrade[] => {
  return tradeIds.toJS().map((id: string) => {
    const trade = trades.get(id)
    if (!trade) {
      throw new Error(`Unexpected trade: ${id}`)
    }
    return trade.toJS()
  })
}

export const findCommonNameByStaticId = (parties: Array<Counterparty | IMember>, staticId?: string) => {
  const member = parties.find((c: Counterparty | IMember) => c.staticId === staticId)
  return member ? get(member, 'x500Name.CN') : `${staticId} not found`
}

export const addBuyerSellerEnrichedData = (
  currentCompanyStaticId: string,
  trades: ITrade[],
  members: Counterparty[] | IMember[]
): ITradeEnriched[] => {
  return trades.map((trade: ITradeEnriched) => ({
    ...trade,
    sellerName: findCommonNameByStaticId(members, trade.seller),
    buyerName: findCommonNameByStaticId(members, trade.buyer),
    tradingRole: currentCompanyStaticId === trade.seller ? TradingRole.SELLER : TradingRole.BUYER
  }))
}

export const selectCounterparty = (trade: ITradeEnriched, tradingRole: TradingRole) => {
  return tradingRole === TradingRole.BUYER ? trade.sellerName : trade.buyerName
}

export const displayCreditRequirement = (creditRequirement?: string): string => {
  const creditRequirementHumanReadableValue =
    CREDIT_REQUIREMENT_DISPLAY_VALUES[creditRequirement ? creditRequirement : '']
  return creditRequirementHumanReadableValue ? creditRequirementHumanReadableValue[0] : UNRECOGNISED_STATUS
}

export const selectNewestLetterOfCreditWithTradeId = (letters: ILetterOfCredit[], tradeId?: string) => {
  return letters
    .filter(l => (l.tradeAndCargoSnapshot ? l.tradeAndCargoSnapshot.trade._id === tradeId : false))
    .sort(
      (a, b) =>
        a.updatedAt && b.updatedAt && typeof a.updatedAt === 'string' && typeof b.updatedAt === 'string'
          ? Date.parse(b.updatedAt) - Date.parse(a.updatedAt)
          : -1
    )[0]
}

export const findFieldFromTradeSchema = (field: string, name: string, schema: any = TRADE_SCHEMA) => {
  return findFieldFromSchema(field, name, schema)
}

export const displayObject = (obj, property, options): string => {
  return JSON.stringify(
    obj,
    (key, value) => {
      if (key) {
        return displayValue(value, key)
      }
      return value
    },
    2
  )
    .replace(/{|}|,|"/g, '')
    .replace(/\\n/g, '\n  ')
    .replace('  \n', '')

  /* output example
    _id: 5c3f0a4d6d25550e4d744bc9
    id: idparcel2
    laycanPeriod:
      startDate: 2018/09/01
      endDate: 2018/09/30
    vesselIMO: 1
    vesselName: Andrej
    loadingPort: Banja luka
    dischargeArea: Sarajevo
    inspector: Boom
    deemedBLDate: 2018/09/01
    quantity: 10
  */
}

export const displayValue = (value, property, options = {}): string => {
  switch (property) {
    case 'maxTolerance':
    case 'minTolerance':
      return displayTolerance(value)
    case 'commodity':
      return displayCommodity(value)
    case 'price':
    case 'amount':
      return formatPrice(value)
    case 'quantity':
      return formatQuantity(value)
    default:
      return handleDefault(value, property, options)
  }
}

const handleDefault = (value, property, options) => {
  if (typeof value === 'boolean') {
    return value ? 'true' : 'false'
  }
  if (isDate(value)) {
    return displayDate(value)
  }
  if (value && typeof value === 'object') {
    return displayObject(value, property, options)
  }
  return value ? value : '-'
}
