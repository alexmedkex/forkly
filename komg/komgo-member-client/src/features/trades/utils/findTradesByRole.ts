import { ITradeEnriched } from '../store/types'
import { TradingRole } from '../constants'

export const findTradesByRole = (trades: ITradeEnriched[], tradingRole: TradingRole, company: string) =>
  trades.filter(trade => trade[tradingRole] === company)
