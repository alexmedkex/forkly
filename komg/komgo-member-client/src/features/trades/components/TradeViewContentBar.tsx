import * as React from 'react'
import { Link } from 'react-router-dom'
import styled from 'styled-components'
import { ContentBar } from '../../../components/content-bar/ContentBar'
import { WithPermissionsProps, WithLicenseCheckProps } from '../../../components'
import { ITradeEnriched } from '../store/types'
import { ILetterOfCredit } from '../../letter-of-credit-legacy/types/ILetterOfCredit'
import { Button, Dropdown } from 'semantic-ui-react'
import { tradeFinanceManager } from '@komgo/permissions'
import { IStandbyLetterOfCredit, IReceivablesDiscountingInfo, RDStatus } from '@komgo/types'
import { productRD } from '@komgo/products'
import { TradingRole } from '../constants'
import {
  getTradeActionsForFinancialInstruments,
  TradeAction,
  getActionPermissions,
  canEditTrade,
  canDeleteTrade
} from '../utils/tradeActionUtils'
import { sentenceCaseWithAcronyms } from '../../../utils/casings'
import _ from 'lodash'
import { Strings } from '../../receivable-discounting-legacy/resources/strings'
import { ILetterOfCreditWithData } from '../../letter-of-credit/store/types'

const StyledHeader = styled.h1`
  display: inline-block;
`

export interface TradeViewContentBarProps extends WithPermissionsProps, WithLicenseCheckProps {
  trade: ITradeEnriched
  role: TradingRole
  legacyLetterOfCredit?: ILetterOfCredit
  standbyLetterOfCredit?: IStandbyLetterOfCredit
  letterOfCredit?: ILetterOfCreditWithData
  rdInfo?: IReceivablesDiscountingInfo
  isStatusFetching: boolean
  hideDeleteButton?: boolean
  hideApplyButtons?: boolean
  onButtonClick: (action?: TradeAction) => void
  onDelete: () => void
}

const renderApplyForFinancing = (actions: TradeAction[], isAuthorized, onButtonClick, hideApplyButtons) => {
  if (hideApplyButtons) {
    return
  }
  const applyActions = actions.filter(action => ![TradeAction.ViewLC, TradeAction.ViewSBLC].includes(action))

  if (applyActions.length > 0) {
    return applyActions.length === 1 ? (
      <Button
        disabled={!_.some(getActionPermissions(applyActions[0]), isAuthorized)}
        primary={true}
        onClick={() => onButtonClick(applyActions[0])}
      >
        {sentenceCaseWithAcronyms(applyActions[0], ['SBLC', 'LC'])}
      </Button>
    ) : (
      <Dropdown
        text="Apply for financing"
        floating={true}
        button={true}
        className="primary"
        direction="left"
        data-test-id="apply-for-financing"
      >
        <Dropdown.Menu>
          {applyActions.map(action => (
            <Dropdown.Item
              key={action}
              data-test-id={`action-${action}`}
              disabled={!_.some(getActionPermissions(action), isAuthorized)}
              content={sentenceCaseWithAcronyms(action, ['SBLC', 'LC'])}
              onClick={() => onButtonClick(action)}
            />
          ))}
        </Dropdown.Menu>
      </Dropdown>
    )
  }
}

const renderViewFinancialInstrumentsActions = (actions: TradeAction[], isAuthorized, onButtonClick) => {
  const viewActions = actions.filter(action => [TradeAction.ViewLC, TradeAction.ViewSBLC].includes(action))
  if (viewActions.length > 0) {
    return viewActions.map(action => (
      <Button
        key={action}
        disabled={!_.some(getActionPermissions(action), isAuthorized)}
        data-test-id={`action-${action}`}
        primary={true}
        onClick={() => onButtonClick(action)}
      >
        {sentenceCaseWithAcronyms(action, ['SBLC', 'LC'])}
      </Button>
    ))
  }
}

const showRDButton = (role: TradingRole, rdInfo?: IReceivablesDiscountingInfo) => {
  const canApply: boolean = !rdInfo || rdInfo.status === RDStatus.PendingRequest
  return role === TradingRole.SELLER && canApply
}

export const TradeViewContentBar: React.SFC<TradeViewContentBarProps> = ({
  trade,
  role,
  legacyLetterOfCredit,
  standbyLetterOfCredit,
  letterOfCredit,
  rdInfo,
  isStatusFetching,
  onButtonClick,
  isLicenseEnabled,
  isAuthorized,
  onDelete,
  hideApplyButtons = false,
  hideDeleteButton = false
}) => {
  const actions = getTradeActionsForFinancialInstruments(
    trade,
    role,
    legacyLetterOfCredit,
    standbyLetterOfCredit,
    letterOfCredit
  )

  return (
    <ContentBar>
      <StyledHeader>Trade details</StyledHeader>
      <div>
        {canEditTrade(trade, rdInfo && rdInfo.status) &&
          isAuthorized(tradeFinanceManager.canCrudTrades) && (
            <Link className="ui button" to={`/trades/${trade!._id}/edit`} data-test-id="edit-trade">
              Edit
            </Link>
          )}
        {canDeleteTrade(trade, legacyLetterOfCredit, standbyLetterOfCredit, letterOfCredit, hideDeleteButton) &&
          isAuthorized(tradeFinanceManager.canCrudTrades) && (
            <Button data-test-id="delete-trade" onClick={onDelete}>
              Delete
            </Button>
          )}

        {!isStatusFetching && renderViewFinancialInstrumentsActions(actions, isAuthorized, onButtonClick)}
        {!isStatusFetching && renderApplyForFinancing(actions, isAuthorized, onButtonClick, hideApplyButtons)}
        {!isStatusFetching &&
          showRDButton(role, rdInfo) && (
            <Button
              primary={true}
              onClick={() => onButtonClick()}
              disabled={!isLicenseEnabled(productRD)}
              data-test-id="apply-for-discounting"
            >
              {Strings.ApplyForDiscountingRiskCoverActionText}
            </Button>
          )}
      </div>
    </ContentBar>
  )
}
