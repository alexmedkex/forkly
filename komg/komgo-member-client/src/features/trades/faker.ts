import { Document, Product } from '../document-management'
import { Products } from '../document-management/constants/Products'
import { fakeDocument } from '../document-management/utils/faker'

export const fakeUploadedDocument = ({ id = '1', name = 'documentName', typeName = 'commercialContract' } = {}) => {
  const mockProduct: Product = { name: 'TRADE FINANCE', id: Products.TradeFinance }
  const mockCategory = { id: 'commercial-contract', name: 'commercial-contract', product: mockProduct }
  const mockType = {
    id: '2',
    name: typeName,
    product: mockProduct,
    category: mockCategory,
    fields: [],
    predefined: true
  }
  const COMPANY_ID = 'societegenerale'
  const doc: Document = fakeDocument({
    id,
    documentId: 'document-id',
    name,
    product: mockProduct,
    category: mockCategory,
    type: mockType,
    owner: { firstName: 'Owner', lastName: 'Owner', companyId: 'company-id' },
    hash: 'hash',
    registrationDate: new Date('2019-02-04T14:47:26.988Z'),
    metadata: [],
    content: undefined,
    sharedWith: [{ counterpartyId: COMPANY_ID, sharedDates: [new Date('2019-02-04T14:47:26.988Z')] }],
    sharedBy: ''
  })
  return doc
}
