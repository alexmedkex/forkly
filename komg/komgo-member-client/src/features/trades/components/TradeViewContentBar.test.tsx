jest.mock('../utils/tradeActionUtils', () => {
  const utils = jest.requireActual('../utils/tradeActionUtils')
  return {
    ...utils,
    getTradeActionsForFinancialInstruments: jest.fn().mockReturnValue([])
  }
})

import * as React from 'react'
import { TradeViewContentBarProps, TradeViewContentBar } from './TradeViewContentBar'
import { fakeTrade } from '../../letter-of-credit-legacy/utils/faker'
import { shallow } from 'enzyme'
import { Button } from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import {
  TradeSource,
  buildFakeReceivablesDiscountingInfo,
  RDStatus,
  CreditRequirements,
  buildFakeTrade
} from '@komgo/types'
import { TradingRole, TRADE_STATUS } from '../constants'
import { TradeAction } from '../utils/tradeActionUtils'
import renderer from 'react-test-renderer'

import { getTradeActionsForFinancialInstruments } from '../utils/tradeActionUtils'
import { render } from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'

const exampleTrade = fakeTrade()
const props: TradeViewContentBarProps = {
  trade: exampleTrade,
  rdInfo: buildFakeReceivablesDiscountingInfo(),
  isLicenseEnabledForCompany: () => true,
  role: TradingRole.BUYER,
  isStatusFetching: false,
  onButtonClick: () => null,
  isAuthorized: () => true,
  isLicenseEnabled: () => true,
  onDelete: () => null
}

const financialInstrumentsActionTest = (action: TradeAction, text: string, enabled: boolean) => {
  it(`should display proper button for ${text}, enabled: ${enabled}`, () => {
    ;(getTradeActionsForFinancialInstruments as jest.Mock).mockReturnValueOnce([action])

    const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} isAuthorized={() => enabled} />)
    const lcButton = content.find(Button)

    expect(lcButton.length).toBe(1)
    expect(lcButton.props().disabled).toBe(!enabled)
    expect(lcButton.dive().text()).toEqual(text)

    expect(getTradeActionsForFinancialInstruments).toHaveBeenCalledWith(
      props.trade,
      TradingRole.BUYER,
      props.legacyLetterOfCredit,
      props.standbyLetterOfCredit,
      props.letterOfCredit
    )
  })
}

describe('TradeViewContentBar', () => {
  describe('should render proper LC buttons', () => {
    it("shoudn't display buttons if no actions", () => {
      ;(getTradeActionsForFinancialInstruments as jest.Mock).mockReturnValueOnce([])

      const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} />)
      expect(content.find(Button).length).toBe(0)
    })

    // added combinations for buttons
    financialInstrumentsActionTest(TradeAction.ApplyForLC, 'Apply for LC', true)
    financialInstrumentsActionTest(TradeAction.ApplyForLC, 'Apply for LC', false)

    financialInstrumentsActionTest(TradeAction.ApplyForSBLC, 'Apply for SBLC', true)
    financialInstrumentsActionTest(TradeAction.ApplyForSBLC, 'Apply for SBLC', false)
    financialInstrumentsActionTest(TradeAction.ViewSBLC, 'View SBLC', true)
    financialInstrumentsActionTest(TradeAction.ViewSBLC, 'View SBLC', false)

    it('should match snapshot', () => {
      ;(getTradeActionsForFinancialInstruments as jest.Mock).mockReturnValueOnce(Object.values(TradeAction))

      expect(renderer.create(<TradeViewContentBar {...props} />).toJSON()).toMatchSnapshot()
    })
  })

  it('should hide delete button if hideDeleteButton is true', () => {
    const { queryByTestId } = render(
      <Router>
        <TradeViewContentBar
          {...props}
          rdInfo={undefined}
          trade={buildFakeTrade({ version: 2 })}
          hideDeleteButton={true}
        />
      </Router>
    )

    expect(queryByTestId('delete-trade')).toBeNull()
  })

  it('should hide apply for financing button if hideApplyButtons is true', () => {
    ;(getTradeActionsForFinancialInstruments as jest.Mock).mockReturnValueOnce([
      TradeAction.ApplyForLC,
      TradeAction.ApplyForSBLC
    ])

    const { queryByTestId } = render(
      <Router>
        <TradeViewContentBar
          {...props}
          rdInfo={undefined}
          trade={buildFakeTrade({ version: 2 })}
          hideApplyButtons={true}
        />
      </Router>
    )

    expect(queryByTestId('apply-for-financing')).toBeNull()
  })

  describe('when we are not the applicant for a trade', () => {
    it('does display the apply for discounting button', () => {
      const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} role={TradingRole.SELLER} />)

      expect(
        content
          .find(Button)
          .first()
          .prop('children')
      ).toEqual('Apply for risk cover / discounting')
    })
  })

  describe('user can edit trade', () => {
    it('should not find link for edit trade if source is VAKT', () => {
      const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} />)
      const link = content.find(Link)

      expect(link.exists()).toBe(false)
    })

    it('should not find link for edit trade if source is KOMGO but user is not authorized', () => {
      const komgoTrade = { ...exampleTrade, source: TradeSource.Komgo }
      const isAuthorized = () => false
      const content = shallow<TradeViewContentBarProps>(
        <TradeViewContentBar {...props} trade={komgoTrade} isAuthorized={isAuthorized} />
      )
      const link = content.find(Link)

      expect(link.exists()).toBe(false)
    })

    it('should find link for edit trade if source is KOMGO and user is authorized', () => {
      const komgoTrade = { ...exampleTrade, source: TradeSource.Komgo }
      const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} trade={komgoTrade} />)

      const link = content.find(Link)

      expect(link.exists()).toBe(true)
    })

    describe('Non-Vakt and Seller Trades', () => {
      beforeEach(() => {
        props.trade = { ...props.trade, source: TradeSource.Komgo }
        props.trade.creditRequirement = CreditRequirements.OpenCredit
        props.trade.tradingRole = TradingRole.SELLER
      })

      it('should find link for edit trade if source is KOMGO and user is authorized and RD Status is QuoteAccepted', () => {
        props.rdInfo.status = RDStatus.QuoteAccepted
        const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} />)

        const link = content.find(Link)

        expect(link.exists()).toBeTruthy()
      })

      it('should find link for edit trade if source is KOMGO and user is authorized and RD Status is PendingRequest', () => {
        props.rdInfo.status = RDStatus.PendingRequest
        const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} />)

        const link = content.find(Link)

        expect(link.exists()).toBeTruthy()
      })

      it('should NOT find link for edit trade if source is KOMGO and user is authorized and RD Status is Requested', () => {
        props.rdInfo.status = RDStatus.Requested
        const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} />)

        const link = content.find(Link)

        expect(link.exists()).toBeFalsy()
      })

      it('should NOT find link for edit trade if source is KOMGO and user is authorized and RD Status is QuoteSubmitted', () => {
        props.rdInfo.status = RDStatus.QuoteSubmitted
        const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} />)

        const link = content.find(Link)

        expect(link.exists()).toBeFalsy()
      })

      it('should find link for edit trade if no RDInfo and TradeStatus is ToBeDiscounted', () => {
        props.rdInfo = undefined
        props.trade = { ...props.trade, status: TRADE_STATUS.ToBeDiscounted }
        const content = shallow<TradeViewContentBarProps>(<TradeViewContentBar {...props} />)

        const link = content.find(Link)

        expect(link.exists()).toBeTruthy()
      })
    })
  })
})
