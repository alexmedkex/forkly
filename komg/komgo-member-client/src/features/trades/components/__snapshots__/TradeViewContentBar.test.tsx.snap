// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TradeViewContentBar should render proper LC buttons should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  height: 50px;
}

.c1 {
  display: inline-block;
}

<header
  className="c0"
>
  <h1
    className="c1"
  >
    Trade details
  </h1>
  <div>
    <button
      className="ui primary button"
      data-test-id="action-VIEW_LC"
      onClick={[Function]}
      role="button"
    >
      View LC
    </button>
    <button
      className="ui primary button"
      data-test-id="action-VIEW_SBLC"
      onClick={[Function]}
      role="button"
    >
      View SBLC
    </button>
    <div
      aria-expanded={false}
      className="ui button floating dropdown primary"
      data-test-id="apply-for-financing"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      role="listbox"
      tabIndex={0}
    >
      <div
        aria-live="polite"
        className="text"
        role="alert"
      >
        Apply for financing
      </div>
      <i
        aria-hidden="true"
        className="dropdown icon"
        onClick={[Function]}
      />
      <div
        className="menu transition left"
      >
        <div
          aria-disabled={true}
          className="disabled item"
          data-test-id="action-EDIT_TRADE"
          onClick={[Function]}
          role="option"
        >
          <span
            className="text"
          >
            Edit trade
          </span>
        </div>
        <div
          aria-disabled={false}
          className="item"
          data-test-id="action-APPLY_FOR_LC"
          onClick={[Function]}
          role="option"
        >
          <span
            className="text"
          >
            Apply for LC
          </span>
        </div>
        <div
          aria-disabled={false}
          className="item"
          data-test-id="action-REAPPLY_FOR_LC"
          onClick={[Function]}
          role="option"
        >
          <span
            className="text"
          >
            Reapply for LC
          </span>
        </div>
        <div
          aria-disabled={false}
          className="item"
          data-test-id="action-APPLY_FOR_SBLC"
          onClick={[Function]}
          role="option"
        >
          <span
            className="text"
          >
            Apply for SBLC
          </span>
        </div>
        <div
          aria-disabled={false}
          className="item"
          data-test-id="action-REAPPLY_FOR_SBLC"
          onClick={[Function]}
          role="option"
        >
          <span
            className="text"
          >
            Reapply for SBLC
          </span>
        </div>
        <div
          aria-disabled={false}
          className="item"
          data-test-id="action-VIEW_LEGACY_SBLC"
          onClick={[Function]}
          role="option"
        >
          <span
            className="text"
          >
            View legacy SBLC
          </span>
        </div>
      </div>
    </div>
  </div>
</header>
`;
